src/sentry/south_migrations/0470_recent_search_backfill.py

```diff
@@ -0,0 +1,1457 @@
+# -*- coding: utf-8 -*-
+import logging
+
+from south.utils import datetime_utils as datetime
+from south.db import db
+from south.v2 import DataMigration
+from django.db import connection
+
+from sentry.utils.hashlib import md5_text
+
+
+db_query = """
+-- Make sure that we only fetch up to 30 searches per user, and filter out any
+-- that already exist
+SELECT organization_id, user_id, 0 as "type", query, date_added
+FROM (
+    SELECT user_id, query, organization_id, date_added, rank() OVER (PARTITION BY user_id ORDER BY date_added DESC) AS pos
+    FROM (
+        -- Group together all saved searches and deduplicate
+        SELECT user_id, query, organization_id, max(date_added) date_added
+        FROM (
+            (
+                -- Get all saved searches that the user explicitly set as default
+                SELECT user_id, query, sp.organization_id, max(ss.date_added) AS date_added
+                FROM sentry_savedsearch_userdefault ssud
+                INNER JOIN sentry_savedsearch ss ON ss.id = ssud.savedsearch_id
+                INNER JOIN sentry_project sp on sp.id = ssud.project_id
+                WHERE ss.query NOT IN ('is:unresolved',
+                                  'is:unresolved is:unassigned',
+                                  'is:unresolved assigned:me',
+                                  'is:unresolved bookmarks:me',
+                                  'is:unresolved age:-24h')
+                GROUP BY user_id, query, sp.organization_id
+            )
+            UNION (
+                -- Also get saved searches that are part of projects they work on
+                SELECT om.user_id, ss.query, sp.organization_id, max(ss.date_added) AS date_added
+                FROM sentry_savedsearch ss
+                INNER JOIN sentry_project sp on sp.id = ss.project_id
+                INNER JOIN sentry_projectteam pt ON pt.project_id = ss.project_id
+                INNER JOIN sentry_organizationmember_teams omt ON omt.team_id = pt.team_id
+                INNER JOIN sentry_organizationmember om ON om.id = omt.organizationmember_id
+                WHERE ss.query NOT IN ('is:unresolved',
+                                  'is:unresolved is:unassigned',
+                                  'is:unresolved assigned:me',
+                                  'is:unresolved bookmarks:me',
+                                  'is:unresolved age:-24h')
+                GROUP BY om.user_id, ss.query, sp.organization_id
+            )
+        ) recent_searches
+        WHERE user_id IS NOT NULL
+        GROUP BY user_id, query, organization_id
+    ) recent_searches
+) AS recent_searches
+WHERE pos <= 30
+and not exists (
+    select *
```

Comment:
Do you need to fetch all columns. Wouldn't fetching only the id work?

Reply:
I think postgres optimizes this for us, but I can just select 1 if you prefer

Reply:
Yeah, just looked at the explain and don't see any difference between these two.

Reply:
Postgres might optimize this away but I know in MySQL if you select all the columns the database will naively build a temporary table with all results and while the query plan won't change you'll be moving more data around inside the database.

Reply:
I did an `EXPLAIN ANALYZE` and didn't see any differences either. Not overly worried about optimizing for mysql. Actually... I don't think our migrations actually run in CI, but I doubt MySQL/sqllite even support `PARTITION BY`. Might just make this migration only run for postgres.